File: app/lib/acp_client.dart
import 'dart:async';
import 'dart:convert';

import 'package:web_socket_channel/web_socket_channel.dart';

class AcpError extends Error {
  final String message;
  AcpError(this.message);
}
class AcpClient {
  WebSocketChannel? _channel;
  int _id = 0;
  final Map<int, Completer<Map<String, dynamic>>> _pending = {};

  StreamSubscription? _sub;

  final StreamController<String> _text = StreamController<String>.broadcast();
  final StreamController<_ToolEvent> _tools = StreamController<_ToolEvent>.broadcast();
  final StreamController<String> _status = StreamController<String>.broadcast();

  String get sessionId {
    if (_channel == null) return '';
    return _sid;
  }
  String _sid = '';

  Stream<String> get text => _text.stream;
  Stream<_ToolEvent> get tools => _tools.stream;
  Stream<String> get status => _status.stream;

  Future<void> connect(String url, {String? secretKey}) async {
    final wsUrl = secretKey == null || secretKey.isEmpty
        ? url
        : '$url?token=${Uri.encodeQueryComponent(secretKey)}';
    _channel = WebSocketChannel.connect(Uri.parse(wsUrl));
    _sub = _channel!.stream.listen(
      (dynamic data) {
        _handle(data);
      },
      onError: (Object e) {
        _status.add('连接错误: $e');
      },
      onDone: () => _status.add('连接已断开'),
    );
    await _send('initialize', {
      'protocolVersion': 1,
      'clientCapabilities': {},
    });
  }

  void _handle(dynamic raw) {
    if (raw is! String) return;
    Map<String, dynamic> msg;
    try {
      msg = jsonDecode(raw) as Map<String, dynamic>;
    } catch (_) {
      return;
    }

    final id = msg['id'];
    if (id is int && _pending.containsKey(id)) {
      _pending.remove(id)?.complete(msg['result'] as Map<String, dynamic>? ?? {});
      return;
    }

    final method = msg['method'] as String?;
    if (method == null) return;

    if (method == 'session/notification') {
      final params = msg['params'] as Map<String, dynamic>? ?? {};
      final update = params['update'] as Map<String, dynamic>?;
      if (update == null) return;
      final uType = update['sessionUpdate'] as String?;
      if (uType == 'agent_message_chunk') {
        final content = (update['content'] as Map<String, dynamic>?)?.['text'];
        if (content is String) _text.add(content);
      } else if (uType == 'tool_call' || uType == 'tool_call_update') {
        _tools.add(_ToolEvent(
          type: uType,
          toolCallId: (update['toolCallId'] as String?) ?? '',
          title: (update['title'] as String?) ??
              (update['toolCall'] as Map<String, dynamic>?)?['title'] as String? ??
              'tool',
          raw: update,
        ));
      }
      } else if (method == 'session_update') {
        final params = msg['params'] as Map<String, dynamic>? ?? {};
        final update = params['update'] as Map<String, dynamic>?;
        if (update != null) {
          _dispatchUpdate(update);
        }
      } else if (method == 'session/update') {
        final params = msg['params'] as Map<String, dynamic>? ?? {};
        final update = params['update'] as Map<String, dynamic>?;
        if (update == null) return;
        _dispatchUpdate(update);
      } else if (method == 'session/request_permission' || method == 'requestPermission') {
        _respondPermission(msg);
      }
    }
  }

  void _dispatchUpdate(Map<String, dynamic> update) {
    final uType = update['sessionUpdate'] as String?;
    if (uType == 'agent_message_chunk') {
      final content = (update['content'] as Map<String, dynamic>?)?.['text'];
      if (content is String) _text.add(content);
    } else if (uType == 'tool_call' || uType == 'tool_call_update') {
      _tools.add(_ToolEvent(
        type: uType,
        toolCallId: (update['toolCallId'] as String?) ?? '',
        title: (update['title'] as String?) ??
            (update['toolCall'] as Map<String, dynamic>?)?['title'] as String? ??
            'tool',
        raw: update,
      ));
    }
  }

  void _respondPermission(Map<String, dynamic> msg) {
    final id = msg['id'];
    if (id == null) return;
    final out = jsonEncode({
      'jsonrpc': '2.0',
      'id': id,
      'result': {
        'outcome': {
          'outcome': 'selected',
          'optionId': 'allow_once',
        },
      },
    });
    _channel?.sink.add(out);
  }

  Future<Map<String, dynamic>> _send(
    String method, [
    Map<String, dynamic>? params,
  ]) async {
    final rid = ++_id;
    final completer = Completer<Map<String, dynamic>>();
    _pending[rid] = completer;
    final out = jsonEncode({
      'jsonrpc': '2.0',
      'id': rid,
      'method': method,
      'params': params,
    });
    _channel?.sink.add(out);
    return completer.future;
  }

  Future<void> newSession({String? cwd}) async {
    final res = await _send('session/new', {
      if (cwd != null) 'cwd': cwd,
      'mcpServers': <Map<String, dynamic>>[],
    });
    _sid = (res['sessionId'] as String?) ?? '';
    _status.add('会话已创建');
  }

  Future<void> prompt(String text) async {
    await _send('session/prompt', {
      'sessionId': _sid,
      'prompt': [
        {
          'type': 'text',
          'text': text,
        }
      ],
    });
  }

  Future<void> dispose() async {
    await _sub?.cancel();
    await _channel?.sink.close();
    _channel = null;
  }
}

class _ToolEvent {
  final String type;
  final String toolCallId;
  final String title;
  final Map<String, dynamic> raw;
  _ToolEvent({required this.type, required this.toolCallId, required this.title, required this.raw});
}

File: web/src/acp-client.ts
// ACP (Agent Client Protocol) WebSocket 客户端。
// 连接 goose serve 的 /acp 端点（JSON-RPC 2.0 over WebSocket）。
// 对外暴露消息流（TextPart / ThoughtPart / ToolPart）和连接状态。

export type AcpStatus = "idle" | "connecting" | "ready" | "error";

export interface ToolPart {
  type: "tool-call";
  toolCallId: string;
  toolName: string;
  status: "in_progress" | "completed" | "failed";
  detail?: string;
}

export type StreamPart =
  | { type: "text"; text: string }
  | { type: "thought"; text: string }
  | ToolPart;

export interface AcpEvents {
  onStatus: (status: AcpStatus, message?: string) => void;
  // 每次调用传入"当前完整快照"（按 messageId 分组后的累积结果），UI 用它刷新消息。
  onSnapshot: (parts: StreamPart[]) => void;
  onRunComplete: () => void;
  onError: (message: string) => void;
}

// 按 messageId 累积 agent 文本；工具调用按 toolCallId 维护状态。
export class AcpClient {
  private ws: WebSocket | null = null;
  private nextId = 0;
  private pending = new Map<
    number,
    { resolve: (v: unknown) => void; reject: (e: unknown) => void; method: string }
  >();
  private sessionId: string | null = null;
  private promptId: number | null = null;
  private ev: AcpEvents;

  // agent 消息累积：messageId -> text
  private textByMessage = new Map<string, string>();
  private thoughtByMessage = new Map<string, string>();
  private toolById = new Map<string, Omit<ToolPart, "type">>();
  private toolOrder: string[] = [];

  constructor(ev: AcpEvents) {
    this.ev = ev;
  }

  get connected() {
    return this.ws?.readyState === WebSocket.OPEN;
  }

  // connect -> initialize -> session/new，完成即 onStatus("ready")
  async connect(url: string): Promise<void> {
    this.ev.onStatus("connecting");
    await new Promise<void>((resolve, reject) => {
      this.ws = new WebSocket(url);
      this.ws.onopen = () => {
        this.initHandshake()
          .then(resolve)
          .catch(reject);
      };
      this.ws.onerror = () => reject(new Error("WebSocket connection failed"));
      this.ws.onclose = () => {
        this.ev.onStatus("idle");
        this.rejectAll(new Error("connection closed"));
      };
      this.ws.onmessage = (e) => this.handleMessage(String(e.data));
    });
    this.ev.onStatus("ready");
  }

  private async initHandshake() {
    await this.request("initialize", {
      protocolVersion: 1,
      clientCapabilities: {},
      clientInfo: { name: "termux-agent-web", version: "0.1" },
    });
    const res = (await this.request("session/new", { cwd: "/tmp", mcpServers: [] })) as {
      sessionId?: string;
    };
    if (!res.sessionId) throw new Error("no sessionId from session/new");
    this.sessionId = res.sessionId;
  }

  // 发送 prompt（fire-and-forget；结果走 session/update 通知 + 最终 result）
  sendPrompt(text: string): void {
    if (!this.sessionId || !this.ws) return;
    // 每条新 prompt 清空累积，避免跨轮串消息
    this.textByMessage.clear();
    this.thoughtByMessage.clear();
    this.toolById.clear();
    this.toolOrder = [];
    this.promptId = this.nextId++;
    const msg = {
      jsonrpc: "2.0",
      id: this.promptId,
      method: "session/prompt",
      params: { sessionId: this.sessionId, prompt: [{ type: "text", text }] },
    };
    this.ws.send(JSON.stringify(msg));
  }

  reset() {
    this.textByMessage.clear();
    this.thoughtByMessage.clear();
    this.toolById.clear();
    this.toolOrder = [];
  }

  private request(method: string, params: unknown): Promise<unknown> {
    return new Promise((resolve, reject) => {
      const id = this.nextId++;
      this.pending.set(id, { resolve, reject, method });
      this.ws?.send(JSON.stringify({ jsonrpc: "2.0", id, method, params }));
    });
  }

  // 默认自动批准权限请求（goose serve 默认 auto 模式）
  respondPermission(requestId: number, approve = true) {
    this.ws?.send(
      JSON.stringify({
        jsonrpc: "2.0",
        id: requestId,
        result: { outcome: { outcome: "selected", optionId: approve ? "allow_once" : "reject_once" } },
      }),
    );
  }

  private handleMessage(raw: string) {
    let msg: any;
    try {
      msg = JSON.parse(raw);
    } catch {
      return;
    }

    // 1. 同步响应（initialize / session/new / session/prompt 最终 result）
    if (msg.id !== undefined && msg.method === undefined) {
      const p = this.pending.get(msg.id);
      if (p) {
        this.pending.delete(msg.id);
        if (msg.error) p.reject(new Error(msg.error.message ?? "rpc error"));
        else p.resolve(msg.result);
      }
      if (this.promptId !== null && msg.id === this.promptId) {
        if (msg.error) this.ev.onError(msg.error.message ?? "prompt failed");
        else this.ev.onRunComplete();
        this.promptId = null;
      }
      return;
    }

    // 2. 服务端主动请求（权限等）
    if (msg.method === "requestPermission") {
      this.respondPermission(msg.id, true);
      return;
    }

    // 3. 通知：session/update
    if (msg.method === "session/update") {
      this.handleUpdate(msg.params?.update ?? msg.params);
    }
  }

  private handleUpdate(update: any) {
    if (!update) return;
    const kind = update.sessionUpdate;

    if (kind === "agent_message_chunk") {
      const mid = String(update.messageId ?? "default");
      const t = update.content?.text ?? "";
      this.textByMessage.set(mid, (this.textByMessage.get(mid) ?? "") + t);
      this.emit();
    } else if (kind === "agent_thought_chunk") {
      const mid = String(update.thoughtMessageId ?? update.messageId ?? "thought");
      const t = update.content?.text ?? "";
      this.thoughtByMessage.set(mid, (this.thoughtByMessage.get(mid) ?? "") + t);
      this.emit();
    } else if (kind === "tool_call") {
      this.upsertTool({
        toolCallId: String(update.toolCallId),
        toolName: update.title ?? "tool",
        status: "in_progress",
      });
    } else if (kind === "tool_call_update") {
      const status: ToolPart["status"] =
        update.status === "completed"
          ? "completed"
          : update.status === "failed" || update.status === "error"
            ? "failed"
            : "in_progress";
      this.upsertTool({
        toolCallId: String(update.toolCallId),
        toolName: update.title ?? this.toolById.get(String(update.toolCallId))?.toolName ?? "tool",
        status,
        detail: update.status,
      });
    }
  }

  private upsertTool(part: Omit<ToolPart, "type">) {
    if (!this.toolById.has(part.toolCallId)) this.toolOrder.push(part.toolCallId);
    this.toolById.set(part.toolCallId, part);
    this.emit();
  }

  // 汇总当前累积状态为完整快照
  private emit() {
    const parts: StreamPart[] = [];
    for (const [mid, text] of this.textByMessage) {
      if (text.trim()) parts.push({ type: "text", text });
    }
    for (const [mid, text] of this.thoughtByMessage) {
      if (text.trim()) parts.push({ type: "thought", text });
    }
    for (const id of this.toolOrder) {
      const p = this.toolById.get(id);
      if (p) parts.push({ ...p, type: "tool-call" as const });
    }
    this.ev.onSnapshot(parts);
  }

  private rejectAll(err: unknown) {
    for (const [, p] of this.pending) p.reject(err);
    this.pending.clear();
  }

  close() {
    this.rejectAll(new Error("closed by client"));
    this.ws?.close();
    this.ws = null;
    this.sessionId = null;
  }
}
